crates/chaintsn/src/transition.rs

```diff
@@ -158,15 +178,28 @@ fn process_l1_checkpoint(
     state: &mut StateCache,
     src_block_mf: &L1BlockManifest,
     signed_ckpt: &SignedCheckpoint,
+    params: &RollupParams,
 ) -> Result<(), TsnError> {
     // TODO verify signature?  it should already have been validated but it
     // doesn't hurt to do it again (just costs)
     // TODO should we verify the proof here too?  the L1 scan proof probably
     // should have done it, but it wouldn't be excessively complicated to
     // re-verify it, we should formally define the answers to these questions
 
+    assert!(
+        verify_signed_checkpoint_sig(signed_ckpt, &params.cred_rule),
+        "Failed to verify signature"
+    );
     let ckpt = signed_ckpt.checkpoint(); // inner data
 
+    let receipt = ckpt.construct_receipt();
+    if !params.proof_publish_mode.allow_empty() {
+        assert!(
+            verify_rollup_groth16_proof_receipt(&receipt, &params.rollup_vk).is_ok(),
+            "posted checkpoint proof verification failed"
+        )
+    }
+
```

Comment:
Should we actually assert here directly? I think it would be better to return Result or and Option. If we have multiple checkpoints, only one of them valid, then returning result/option would allow us to try to verify checkpoint until it passes. If none pass then we can probably panic or assert.

Reply:
I also think we should check if the proof is the expected proof i.e. proof we are expecting for the next epoch.